File: 21-merge-two-sorted-lists/merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* ansHead = new ListNode();
        ListNode* ansTail = ansHead;

        while(list1 != NULL && list2 != NULL) {
            int val1 = list1->val;
            int val2 = list2->val;

            if(val1 < val2) { // take from 1
                ansTail->next = list1;
                list1 = list1->next;
            } else { // take from 2
                ansTail->next = list2;
                list2 = list2->next;     
            }
            ansTail = ansTail->next;
        }

        if(list1 != NULL) ansTail->next = list1;
        else if(list2 != NULL) ansTail->next = list2;

        return ansHead->next;
    }
};

File: 21-merge-two-sorted-lists/merge-two-sorted-lists.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
    if(list1 == null) {
        return list2;
    }
    if(list2 == null) {
        return list1;
    }
    let head = new ListNode(0);
    let curr = head;
    while(list1 != null && list2 != null) {
        if(list1.val < list2.val) {
            curr.next = list1;
            list1 = list1.next;
            curr = curr.next;
        } else {
            curr.next = list2;
            list2 = list2.next;
            curr = curr.next;
        }
    }
    if(list1 != null) {
        curr.next = list1;
    } else if(list2 != null) {
        curr.next = list2;
    }
    return head.next;
};